File: DSA/LC/242_valid_anagram.java
import java.util.HashMap;

class Solution {
  public static boolean isAnagram(String s, String t) {
    // char[] s_arr = s.toCharArray();
    // char[] t_arr = t.toCharArray();
    // Arrays.sort(s_arr);
    // Arrays.sort(t_arr);
    // s = new String(s_arr);
    // t = new String(t_arr);
    // return s.equals()
    HashMap<Character, Integer> s_map = new HashMap<>();
    HashMap<Character, Integer> t_map = new HashMap<>();

    if (s.length() != t.length()) {
      return false;
    }

    for (char c : s.toCharArray()) {
      s_map.put(c, s_map.getOrDefault(c, 0) + 1);
    }

    for (char c : t.toCharArray()) {
      t_map.put(c, t_map.getOrDefault(c, 0) + 1);
    }

    if (t_map.size() != s_map.size()) {
      return false;
    }

    for (Character c : s_map.keySet()) {
      if (!t_map.containsKey(c) || !t_map.get(c).equals(s_map.get(c))) {
        return false;
      }
    }
    return true;
  }
}

File: DSA/LC/242_valid_anagram.js
class Solution {
  /**
   * @param {string} s
   * @param {string} t
   * @return {boolean}
   */
  isAnagram(s, t) {
    // return s.toLowerCase().split('').sort().join('') === t.toLowerCase().split('').sort().join('');
    const s_map = {};
    const t_map = {};

    if (s.length != t.length) {
      return false;
    }

    s.split("").forEach((c) => {
      s_map[c] = (s_map[c] || 0) + 1;
    });

    t.split("").forEach((c) => {
      t_map[c] = (t_map[c] || 0) + 1;
    });

    const s_keys = Object.keys(s_map);
    const t_keys = Object.keys(t_map);

    if (s_keys.length != t_keys.length) {
      return false;
    }

    let outcome = true;

    s_keys.forEach((k) => {
      if (t_map[k] !== s_map[k]) {
        outcome = false;
        return false;
      }
    });

    return outcome;
  }
}
